internal/pkg/service/stream/storage/node/coordinator/fileimport/operator.go

```diff
@@ -0,0 +1,272 @@
+	})
+
+	// Mirror files
+	{
+		o.files = etcdop.SetupMirrorMap[model.File, model.FileKey, *fileData](
+			d.StorageRepository().File().GetAllInLevelAndWatch(ctx, model.LevelStaging, etcd.WithPrevKV()),
+			func(_ string, file model.File) model.FileKey {
+				return file.FileKey
+			},
+			func(_ string, file model.File, rawValue *op.KeyValue, oldValue **fileData) *fileData {
+				out := &fileData{
+					FileKey: file.FileKey,
+					State:   file.State,
+					Retry:   file.Retryable,
+					IsEmpty: file.StagingStorage.IsEmpty,
+					File: plugin.File{
+						FileKey:  file.FileKey,
+						SinkKey:  file.SinkKey,
+						Provider: file.TargetStorage.Provider,
+					},
+				}
+
+				// Keep the same lock, to prevent parallel processing of the same file.
+				// No modification from another code is expected, but just to be sure.
+				if oldValue != nil {
+					out.Lock = (*oldValue).Lock
+				} else {
+					out.Lock = &sync.Mutex{}
+				}
+
+				return out
+			},
+		).
+			// Check only files owned by the node
+			WithFilter(func(event etcdop.WatchEvent[model.File]) bool {
+				return o.distribution.MustCheckIsOwner(event.Value.SourceKey.String())
+			}).
+			BuildMirror()
+		if err = <-o.files.StartMirroring(ctx, wg, o.logger); err != nil {
+			return err
+		}
+	}
+
+	// Restarts stream on distribution change
+	{
+		wg.Add(1)
+		listener := o.distribution.OnChangeListener()
+
+		go func() {
+			defer wg.Done()
+			defer listener.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case events := <-listener.C:
+					o.files.Restart(errors.Errorf("distribution changed: %s", events.Messages()))
+				}
+			}
+		}()
+	}
+
+	// Start importing files check ticker
+	{
+		wg.Add(1)
+		ticker := d.Clock().Ticker(o.config.FileImportCheckInterval.Duration())
+
+		go func() {
+			defer wg.Done()
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					o.checkFiles(ctx, wg)
+				}
+			}
+		}()
+	}
+
+	return nil
+}
+
+func (o *operator) checkFiles(ctx context.Context, wg *sync.WaitGroup) {
+	o.logger.Debugf(ctx, "checking files in the importing state")
+
+	o.files.ForEach(func(_ model.FileKey, file *fileData) (stop bool) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			o.checkFile(ctx, file)
+		}()
+		return false
+	})
+}
+
+func (o *operator) checkFile(ctx context.Context, file *fileData) {
+	// Prevent multiple checks of the same file
+	if !file.Lock.TryLock() {
+		return
+	}
+	defer file.Lock.Unlock()
+
+	// File has been modified by some previous check, but we haven't received an updated version from etcd yet
+	if file.Processed {
+		return
+	}
+
+	if !file.Retry.Allowed(o.clock.Now()) {
+		return
+	}
+
+	switch file.State {
+	case model.FileImporting:
+		o.importFile(ctx, file)
+	default:
+		// nop
+	}
+}
+
+func (o *operator) importFile(ctx context.Context, file *fileData) {
+	o.logger.Info(ctx, "importing file")
+
+	var err error
```

Comment:
Here should be a `ctx` with th `FileImportTimeout` - the timeout applies to the whole operation, not only to the state switch.